web_sdk/web_test_utils/lib/skia_client.dart

```diff
@@ -16,6 +16,13 @@ const String _kGoldctlKey = 'GOLDCTL';
 const String _skiaGoldHost = 'https://flutter-engine-gold.skia.org';
 const String _instance = 'flutter-engine';
 
+/// The percentage of accepted pixels to be wrong.
+///
+/// This should be a double between 0.0 and 1.0. A value of 0.0 means we don't
+/// accept any pixel to be different. A value of 1.0 means we accept 100% of
+/// pixels to be different.
+const double kMaxDifferentPixelsRate = 0.1;
```

Comment:
I think we should use an absolute number of pixels and not depend on the screenshot size. I think I made a mistake when I decided to make it a percentage in the past.

Reply:
Having said that, now that we have a per-pixel threshold, I feel like maybe this value should either be 0% or 100% :)

Reply:
This number was also picked based on false negatives that I audited :)

The maximum number of pixels that were different in a false negative was around 6-7%. So I added some buffer at 10%.

We can't set this number to zero because that defeats the purpose of using fuzzy matching. Zero basically means: fail if any single pixel is different.

We only expect a subset of pixels to be different e.g. borders, text characters, shadows, etc. So I think going a more conservative approach at 10% is better than going full 100%.

That said, all of this is subject to change as we learn more in the future.

Reply:
SGTM. Do I understand correctly that the final condition to pass a golden test is:

- at least 90% of pixels are 100% identical to golden pixels, _and_
- no pixel is more than `pixelDeltaThreshold` different that the corresponding golden pixel.

If either of the above conditions do not hold the test fails. Is that correct?

Reply:
Your understanding is correct. Here's the pseudo-algorithm I found in the fuzzy matching design doc:
```python
# Iterate over all pixels.
for x in range(expectedImg.width):
  for y in range(expectedImg.height):
    # Number of different pixels must be below the
    # given threshold.
    if expectedImg[x][y] != actualImg[x][y]:
      numDiffPixels++
    if numDiffPixels > maxDiffPixels:
      return False

    # Pixel-wise differences must be below the
    # given threshold.
    p = expectedImg[x][y], q = actualImg[x][y]
    delta = abs(p.r - q.r)
          + abs(p.g - q.g)
          + abs(p.b - q.b)
          + abs(p.a - q.a)
    if delta > pixelThreshold:
      return False
return True
```